### PR TITLE
MAJOR refactoring to use generated decoders

### PR DESCRIPTION
**Pull request by cloudify**
Also simplifies handling of responses from upstream APIs

### CODE REVIEW

**Comment by cloudify** [comment]
This pull request **introduces 2 alerts** when merging 46e3cc56749c8aea996e72cbf19d83b3d6925edd into 9567485c4a636b921e9607a941ba4fc25a44ad46 - [view on LGTM.com](https://lgtm.com/projects/g/teamdigitale/io-backend/rev/pr-adcd27e33a84b5f474c0c65980cda8c402e750e9)

**new alerts:**

* 2 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by cloudify** [comment]
This pull request **introduces 1 alert** when merging e5261250c6d6fa086fd55f1dde857cc3565106a3 into 9567485c4a636b921e9607a941ba4fc25a44ad46 - [view on LGTM.com](https://lgtm.com/projects/g/teamdigitale/io-backend/rev/pr-41e32c08c0e631b63da686b285430e46720e8a5a)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a Pivotal story at the beginning of the PR title (see below).
  </td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    

New dependencies added: <a href='https://www.npmjs.com/package/fp-ts'><code>fp-ts</code></a> and <a href='https://www.npmjs.com/package/typescript'><code>typescript</code></a>.


<h2><a href="https://github.com/gcanti/fp-ts">fp-ts</a></h2>
<p>Author: Giulio Canti</p>
<p>Description: Functional programming in TypeScript</p>
<p>Homepage: <a href="https://github.com/gcanti/fp-ts">https://github.com/gcanti/fp-ts</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 2 years ago</td></tr><tr><td>Last Updated</td><td>4 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>89</td></tr><tr><td>Direct Dependencies</td><td></td></tr><tr><td>Keywords</td><td>typescript, algebraic-data-types and functional-programming</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

<h3 align="center">
  <a href="https://gcanti.github.io/fp-ts/">
    
  </a>
</h3>

<p align="center">
Functional programming in TypeScript
</p>

<p align="center">
  <a href="https://travis-ci.org/gcanti/fp-ts">
    build status
  </a>
  <a href="https://david-dm.org/gcanti/fp-ts">
    dependency status
  </a>
  <a href="https://www.npmjs.com/package/fp-ts">
    npm downloads
  </a>
</p>

# Typed functional programming in TypeScript

`fp-ts` is a library for _typed functional programming_ in TypeScript.

`fp-ts` aims to allow developers to use _popular patterns and abstractions_ that are available in most functional languages. For this, it includes the most popular data types, type classes and abstractions such as [Option](https://gcanti.github.io/fp-ts/modules/Option.ts), [Either](https://gcanti.github.io/fp-ts/modules/Either.ts), [IO](https://gcanti.github.io/fp-ts/modules/IO.ts), [Task](https://gcanti.github.io/fp-ts/modules/Task.ts), [Functor](https://gcanti.github.io/fp-ts/modules/Functor.ts), [Applicative](https://gcanti.github.io/fp-ts/modules/Applicative.ts), [Monad](https://gcanti.github.io/fp-ts/modules/Monad.ts) to empower users to write pure FP apps and libraries built atop higher order abstractions.

A distinctive feature of `fp-ts` with respect to other functional libraries is its implementation of [Higher Kinded Types](<https://en.wikipedia.org/wiki/Kind_(type_theory)>), which TypeScript doesn't support natively.

**Inspired by**

- [Haskell](https://haskell-lang.org)
- [PureScript](http://www.purescript.org)
- [Scala](https://www.scala-lang.org/)




**Table of contents**

- [Installation](#installation)
  - [TypeScript compatibility](#typescript-compatibility)
- [Documentation](#documentation)
- [Development](#development)
- [License](#license)



# Installation

To install the stable version:

```
npm install fp-ts
```

Make sure to always have a single version of `fp-ts` installed in your project. Multiple versions are known to cause `tsc` to hang during compilation. You can check the versions currently installed using `npm ls fp-ts` (make sure there's a single version and all the others are marked as `deduped`).

## Consuming fp-ts

Most examples will use the following import syntax:

```
import { Option, some, none } from 'fp-ts/lib/Option'
```

This will give you the widest support across tools, as you will be importing CommonJS modules.

If you use a bundler such as webpack or Rollup that supports tree-shaking, you can take advantage of this by opting to import ECMAScript modules instead:

```
import { Option, some, none } from 'fp-ts/es6/Option'
```

Note that there are caveats such as some tools (e.g. Jest) not supporting ES6 module syntax natively yet.

## TypeScript compatibility

**Strictness** – This library is conceived, tested and is supposed to be consumed by TypeScript with the `strict` flag turned on.

| `fp-ts` version | required `typescript` version |
| --------------- | ----------------------------- |
| 1.15.x+         | 3.1+                          |
| <= 1.14.4       | 2.8+ (\*)                     |

(\*) If you are running `< typescript@3.0.1` you have to polyfill the `unknown` type. You can use [unknown-ts](https://github.com/gcanti/unknown-ts) as a polyfill.

# Documentation

- [Docs](https://gcanti.github.io/fp-ts)
- [Tutorial](https://gcanti.github.io/fp-ts/basics/)
- [Ecosystem](https://gcanti.github.io/fp-ts/introduction/ecosystem)
- [API Reference](https://gcanti.github.io/fp-ts/modules/)

# Development

- [Code conventions](https://gcanti.github.io/fp-ts/introduction/code-conventions)

# License

The MIT License (MIT)


</details>




<h2><a href="https://www.typescriptlang.org/">typescript</a></h2>
<p>Author: Microsoft Corp.</p>
<p>Description: TypeScript is a language for application scale JavaScript development</p>
<p>Homepage: <a href="https://www.typescriptlang.org/">https://www.typescriptlang.org/</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 6 years ago</td></tr><tr><td>Last Updated</td><td>about 2 hours ago</td></tr><tr><td>License</td><td>Apache-2.0</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>1272</td></tr><tr><td>Direct Dependencies</td><td></td></tr><tr><td>Keywords</td><td>TypeScript, Microsoft, compiler, language and javascript</td></tr>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by cloudify** [comment]
This pull request **introduces 2 alerts** when merging 45232b8e533948bfa8f5f2df1b9f27d9a1141baf into 9567485c4a636b921e9607a941ba4fc25a44ad46 - [view on LGTM.com](https://lgtm.com/projects/g/teamdigitale/io-backend/rev/pr-c8ed07c5f04d35838d6604a4bd92b9719f0616a8)

**new alerts:**

* 2 for Incomplete string escaping or encoding

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/controllers/sessionController.ts [hunk 1]
```diff
@@ -3,38 +3,24 @@
...
-    return ResponseSuccessJson({
-      spidLevel: user.spid_level,
-      walletToken: user.wallet_token
-    });
-  }
+export default class SessionController {
+  public readonly getSessionState = (
+    req: express.Request
+  ): Promise<IResponseErrorValidation | IResponseSuccessJson<PublicSession>> =>
+    withUserFromRequest(req, async user =>
```
Comment (gunzip) [inline src/controllers/sessionController.ts]: nit: if you curry `withUserFromRequest(callback)(request)` you can use directly
`getSessionState = withUserFromRequest(callback)`

File: src/utils/responses.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Transforms async failures into internal errors
+ */
+export const withCatchAsInternalError = <T>(
+  f: () => Promise<T>,
+  message: string = "Exception while calling upstream API (likely a timeout)."
+) =>
+  f().catch(_ => {
+    // tslint:disable-next-line:no-console
+    console.error(_);
```
Comment (gunzip) [inline src/utils/responses.ts]: is this on purpose ?

Comment (cloudify) [reply src/utils/responses.ts]: good catch, no I think I added it before adding the logging of the responses

File: src/controllers/sessionController.ts (hunk 1 above)
Comment (cloudify) [reply src/controllers/sessionController.ts]: care to send a PR with this improvement? 👍

**Pull request merged**